velox/functions/lib/Re2Functions.cpp

```diff
@@ -805,71 +880,95 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
-  vector_size_t patternLength = pattern.size();
-  vector_size_t i = 0;
-  // Index of the first % or _ character.
-  vector_size_t wildcardStart = -1;
-  // Index of the first character that is not % and not _.
-  vector_size_t fixedPatternStart = -1;
-  // Total number of % characters.
-  vector_size_t anyCharacterWildcardCount = 0;
-  // Total number of _ characters.
-  vector_size_t singleCharacterWildcardCount = 0;
-  auto patternStr = pattern.data();
+PatternParameters determinePatternKind(const StringView& pattern) {
+  auto i = 0;
+  const auto patternLength = pattern.size();
+  const auto patternString = pattern.str();
+  auto wildcardStart = 0;
+  auto wildcardCount = 0;
+  auto anyCharacterWildcardCount = 0;
+  auto singleCharacterWildcardCount = 0;
+  auto fixedPatternStart = 0;
+  auto fixedPatternCount = 0;
+  std::vector<std::string> fixedPatterns;
 
   while (i < patternLength) {
-    if (patternStr[i] == '%' || patternStr[i] == '_') {
-      // Ensures that pattern has a single contiguous stream of wildcard
-      // characters.
-      if (wildcardStart != -1) {
-        return std::make_pair(PatternKind::kGeneric, 0);
-      }
+    if (patternString[i] == '%' || patternString[i] == '_') {
       // Look till the last contiguous wildcard character, starting from this
       // index, is found, or the end of pattern is reached.
       wildcardStart = i;
       while (i < patternLength &&
-             (patternStr[i] == '%' || patternStr[i] == '_')) {
-        singleCharacterWildcardCount += (patternStr[i] == '_');
-        anyCharacterWildcardCount += (patternStr[i] == '%');
+             (patternString[i] == '%' || patternString[i] == '_')) {
+        singleCharacterWildcardCount += (patternString[i] == '_');
+        anyCharacterWildcardCount += (patternString[i] == '%');
         i++;
       }
-    } else {
-      // Ensure that pattern has a single fixed pattern.
-      if (fixedPatternStart != -1) {
-        return std::make_pair(PatternKind::kGeneric, 0);
+      if (!singleCharacterWildcardCount) {
+        wildcardCount += (anyCharacterWildcardCount > 0);
       }
+    } else {
       // Look till the end of fixed pattern, starting from this index, is found,
       // or the end of pattern is reached.
       fixedPatternStart = i;
       while (i < patternLength &&
-             (patternStr[i] != '%' && patternStr[i] != '_')) {
+             (patternString[i] != '%' && patternString[i] != '_')) {
+        i++;
+      }
+      if (i != fixedPatternStart) {
+        fixedPatternCount++;
+        fixedPatterns.emplace_back(
+            patternString.substr(fixedPatternStart, i - fixedPatternStart));
+      } else {
         i++;
       }
     }
   }
 
   // Pattern contains wildcard characters only.
-  if (fixedPatternStart == -1) {
-    if (!anyCharacterWildcardCount) {
-      return {PatternKind::kExactlyN, singleCharacterWildcardCount};
+  if (fixedPatternCount <= 1 && wildcardCount <= 1) {
```

Comment:
These ifs can be rewritten in a more readable manner :
if (fixedPatternCount == 0 && wildCardCount <= 1) {
  return the 2 possibilities
}

if (fixedPatternCount == 1 && singleCharacterWildcardCount) {
...
}

Avoid use of else , if all previous paths return. You can simply write a series of if condition statements.

Reply:
Thanks for the suggestion, simplified the checks to make them more readable.